include/deal.II/base/mpi_remote_point_evaluation.h

```diff
@@ -0,0 +1,514 @@
+       *   vertex) that is shared by multiple cells or a point is outside of the
+       *   computational domain.
+       */
+      template <typename T>
+      void
+      evaluate_and_process(
+        std::vector<T> &output,
+        std::vector<T> &buffer,
+        const std::function<void(const ArrayView<T> &, const CellData &)> &fu)
+        const;
+
+      /**
+       * This method is the inverse of the method evaluate_and_process(). It
+       * makes data at the points and provided by @p input available in the
+       * function @p fu.
+       */
+      template <typename T>
+      void
+      process_and_evaluate(
+        const std::vector<T> &input,
+        std::vector<T> &      buffer,
+        const std::function<void(const ArrayView<const T> &, const CellData &)>
+          &fu) const;
+
+      /**
+       * Return a CRS-like data structure to determine the position of the
+       * result corresponding a point and the amount.
+       */
+      const std::vector<unsigned int> &
+      get_point_ptrs() const;
+
+      /**
+       * Return if points and cells have a one-to-one relation. This is not the
+       * case if a points is not owned by any cell (the point is outside of the
+       * domain) or if multiple cells own the point (the point is positioned
+       * on a geometric entity shared by neighboring cells).
+       */
+      bool
+      is_map_unique() const;
+
+      /**
+       * Return the Triangulation object used during reinit().
+       */
+      const Triangulation<dim, spacedim> &
+      get_triangulation() const;
+
+      /**
+       * Return the Mapping object used during reinit().
+       */
+      const Mapping<dim, spacedim> &
+      get_mapping() const;
+
+      /**
+       * Return if the internal data structures have been set up and if yes
+       * if they are still valid (and have not been invalidated due to changes
+       * of the Triangulation).
+       */
+      bool
+      is_ready() const;
+
+    private:
+      /**
+       * Tolerance to be used while determining the surrounding cells of a
+       * point.
+       */
+      const double tolerance;
+
+      /**
+       * Storage for the status of the triangulation signal.
+       */
+      boost::signals2::connection tria_signal;
+
+      /**
+       * Flag indicating if the reinit() function has been called and if yes
+       * the triangulation has not been modified since then (potentially
+       * invalidating the communication pattern).
+       */
+      bool ready_flag;
+
+      /**
+       * Reference to the Triangulation object used during reinit().
+       */
+      SmartPointer<const Triangulation<dim, spacedim>> tria;
+
+      /**
+       * Reference to the Mapping object used during reinit().
+       */
+      SmartPointer<const Mapping<dim, spacedim>> mapping;
+
+      /**
+       * MPI communicator of the triangulation.
+       */
+      MPI_Comm comm;
```

Comment:
Do we still need this now that it is reasonably simple to query the communication from the grid? Or do we envision cases where we might use another communicator?